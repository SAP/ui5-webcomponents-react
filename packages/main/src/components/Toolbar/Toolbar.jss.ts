import { CssSizeVariables, ThemingParameters } from '@ui5/webcomponents-react-base';

export const styles = {
  outerContainer: {
    boxSizing: 'border-box',
    width: '100%',
    maxWidth: '100%',
    height: CssSizeVariables.sapWcrToolbarHeight,
    position: 'relative',
    overflow: 'hidden',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottom: `solid 0.0625rem ${ThemingParameters.sapGroup_TitleBorderColor}`
  },
  hasOverflow: {
    '& $toolbar': {
      maxWidth: 'calc(100% - 44px)'
    }
  },
  clear: {
    borderBottom: 'none'
  },
  active: {
    cursor: 'pointer',
    '&:active': {
      backgroundColor: ThemingParameters.sapActiveColor
    },
    '&:hover': {
      backgroundColor: ThemingParameters.sapList_Hover_Background
    }
  },
  info: {
    // This color is most similar to darken(@sapUiBaseColor, 10)
    backgroundColor: ThemingParameters.sapList_HeaderBorderColor,
    '&$active': {
      backgroundColor: ThemingParameters.sapInfobar_Active_Background,
      '&:active': {
        backgroundColor: `${ThemingParameters.sapInfobar_Active_Background}`
      },
      '&:hover': {
        backgroundColor: ThemingParameters.sapInfobar_Hover_Background
      }
    }
  },
  solid: {
    backgroundColor: ThemingParameters.sapBackgroundColor
  },
  transparent: {
    backgroundColor: ThemingParameters.sapToolbar_Background
  },
  toolbar: {
    width: '100%',
    '& >:first-child:not(.spacer)': {
      margin: '0 0.25rem 0 0'
    },
    '& >:last-child:not(.spacer)': {
      margin: '0 0.5rem 0 0.25rem'
    },
    '& > *:not(first-child):not(last-child):not(.spacer)': {
      margin: '0 0.25rem'
    },
    display: 'flex',
    alignItems: 'center',
    maxWidth: '100%'
  },
  overflowButtonContainer: {
    display: 'flex',
    marginRight: '0.5rem'
  },
  popover: {
    '&[ui5-popover]::part(content)': {
      padding: 0
    }
  },
  popoverPhone: {
    width: 'calc(100% - 10px)',
    maxWidth: 'calc(100% - 10px)',
    left: '5px !important'
  },
  popoverContent: {
    padding: CssSizeVariables.sapWcrToolbarPopoverContentPadding,
    display: 'flex',
    flexDirection: 'column',
    '& [ui5-button]': {
      marginBottom: '0.25rem'
    },
    '& [ui5-button]::part(button)': {
      justifyContent: 'flex-start'
    },
    '& :last-child': {
      marginBottom: 0
    }
  },
  childContainer: { display: 'flex' }
};
