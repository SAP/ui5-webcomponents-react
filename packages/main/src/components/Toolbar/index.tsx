import { createUseStyles } from 'react-jss';
import { useConsolidatedRef, useI18nBundle, useIsomorphicLayoutEffect } from '@ui5/webcomponents-react-base/dist/hooks';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/dist/StyleClassHelper';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/dist/Utils';
import { SHOW_MORE } from '@ui5/webcomponents-react/dist/assets/i18n/i18n-defaults';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import { ToolbarDesign } from '@ui5/webcomponents-react/dist/ToolbarDesign';
import { ToolbarStyle } from '@ui5/webcomponents-react/dist/ToolbarStyle';
import React, {
  cloneElement,
  createRef,
  forwardRef,
  ReactElement,
  ReactFragment,
  ReactNode,
  ReactNodeArray,
  Ref,
  RefObject,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from 'react';
import { OverflowPopover } from './OverflowPopover';
import { styles } from './Toolbar.jss';

const useStyles = createUseStyles(styles, { name: 'Toolbar' });

export interface ToolbarPropTypes extends Omit<CommonProps, 'onClick'> {
  /**
   * Defines the content of the `Toolbar`.
   */
  children?: ReactNode | ReactNodeArray | ReactFragment;
  /**
   * Defines the visual style of the `Toolbar`.<br />
   * <b>Note:</b> The visual styles are theme-dependent.
   */
  toolbarStyle?: ToolbarStyle | keyof typeof ToolbarStyle;
  /**
   * Defines the `Toolbar` design.<br />
   * <b>Note:</b> Design settings are theme-dependent.
   */
  design?: ToolbarDesign | keyof typeof ToolbarDesign;
  /**
   * Indicates that the whole `Toolbar` is clickable. The Press event is fired only if `active` is set to true.
   */
  active?: boolean;
  /**
   * Sets the components outer HTML tag.
   *
   * __Note:__ For TypeScript the types of `ref` are bound to the default tag name, if you change it you are responsible to set the respective types yourself.
   */
  as?: keyof HTMLElementTagNameMap;
  /**
   * Fired when the user clicks on the `Toolbar`, if the `active` prop is set to "true".
   */
  onClick?: (event: CustomEvent) => void;
}

/**
 * Horizontal container most commonly used to display buttons, labels, selects and various other input controls.
 *
 * The content of the `Toolbar` moves into the overflow area from right to left when the available space is not enough in the visible area of the container.
 * It can be accessed by the user through the overflow button that opens it in a popover.
 */
const Toolbar = forwardRef((props: ToolbarPropTypes, ref: Ref<HTMLDivElement>) => {
  const { children, toolbarStyle, design, active, style, tooltip, className, onClick, slot, as, ...rest } = props;
  const classes = useStyles();
  const outerContainer: RefObject<HTMLDivElement> = useConsolidatedRef(ref);
  const controlMetaData = useRef([]);
  const [lastVisibleIndex, setLastVisibleIndex] = useState<number>(null);

  const i18nBundle = useI18nBundle('@ui5/webcomponents-react');

  const toolbarClasses = StyleClassHelper.of(classes.outerContainer);
  if (toolbarStyle === ToolbarStyle.Clear) {
    toolbarClasses.put(classes.clear);
  }
  if (active) {
    toolbarClasses.put(classes.active);
  }
  switch (design) {
    case ToolbarDesign.Solid:
      toolbarClasses.put(classes.solid);
      break;
    case ToolbarDesign.Transparent:
      toolbarClasses.put(classes.transparent);
      break;
    case ToolbarDesign.Info:
      toolbarClasses.put(classes.info);
      break;
    default:
      break;
  }
  toolbarClasses.putIfPresent(className);

  const childrenWithRef = useMemo(() => {
    controlMetaData.current = [];

    const refactoredChildren = React.Children.toArray(children).map((child, index) => {
      if ((child as ReactElement).type === React.Fragment) {
        return (child as ReactElement).props.children.filter(Boolean).map((item, itemIndex: number) => {
          return cloneElement(item, { key: `.${index}:${itemIndex}` });
        });
      }
      return child;
    });

    return refactoredChildren.flat().map((item: ReactElement, index) => {
      const itemRef: RefObject<HTMLDivElement> = createRef();

      controlMetaData.current.push({
        ref: itemRef
      });

      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      if ((item?.type as any)?.displayName === 'ToolbarSpacer') {
        return item;
      }
      return (
        <div ref={itemRef} key={index} className={classes.childContainer} data-component-name="ToolbarChildContainer">
          {item}
        </div>
      );
    });
  }, [children, controlMetaData, classes.childContainer]);

  const overflowNeeded =
    (lastVisibleIndex || lastVisibleIndex === 0) && React.Children.count(childrenWithRef) !== lastVisibleIndex + 1;

  toolbarClasses.putIfPresent(overflowNeeded && classes.hasOverflow);

  const requestAnimationFrameRef = useRef<undefined | number>();
  const calculateVisibleItems = useCallback(() => {
    const OVERFLOW_BUTTON_WIDTH = 32 + 8;
    requestAnimationFrameRef.current = requestAnimationFrame(() => {
      if (!outerContainer.current) return;
      const availableWidth = outerContainer.current.getBoundingClientRect().width;
      let consumedWidth = 0;
      let lastIndex = null;
      let lastFitWidth = 0;

      if (availableWidth - OVERFLOW_BUTTON_WIDTH <= 0) {
        lastIndex = -1;
        lastFitWidth = 0;
      } else {
        controlMetaData.current.forEach((item, index) => {
          const currentMeta = controlMetaData.current[index] as { ref: RefObject<HTMLElement> };
          if (currentMeta && currentMeta.ref && currentMeta.ref.current) {
            let nextWidth = currentMeta.ref.current.getBoundingClientRect().width;
            nextWidth += index === 0 || index === controlMetaData.current.length - 1 ? 4 : 8; // first & last element = padding: 4px
            if (index === controlMetaData.current.length - 1) {
              if (consumedWidth + nextWidth <= availableWidth - 8) {
                lastIndex = index;
                lastFitWidth = consumedWidth + nextWidth;
              }
            } else {
              if (consumedWidth + nextWidth <= availableWidth - OVERFLOW_BUTTON_WIDTH) {
                lastIndex = index;
                lastFitWidth = consumedWidth + nextWidth;
              }
              if (
                consumedWidth < availableWidth - OVERFLOW_BUTTON_WIDTH &&
                consumedWidth + nextWidth >= availableWidth - OVERFLOW_BUTTON_WIDTH
              ) {
                lastIndex = index - 1;
                lastFitWidth = 0;
              }
            }
            consumedWidth += nextWidth;
          }
        });
      }
      setLastVisibleIndex(lastIndex);
    });
  }, [outerContainer.current, controlMetaData.current, setLastVisibleIndex, childrenWithRef, overflowNeeded]);

  const observer = useRef(new ResizeObserver(calculateVisibleItems));

  useEffect(() => {
    if (outerContainer.current) {
      observer.current.observe(outerContainer.current);
    }
    return () => {
      cancelAnimationFrame(requestAnimationFrameRef.current);
      observer.current.disconnect();
    };
  }, [outerContainer.current]);

  useIsomorphicLayoutEffect(() => {
    calculateVisibleItems();
  }, [calculateVisibleItems]);

  const handleToolbarClick = useCallback(
    (e) => {
      if (active && typeof onClick === 'function') {
        onClick(enrichEventWithDetails(e));
      }
    },
    [onClick, active]
  );

  const CustomTag = as as React.ElementType;
  return (
    <CustomTag
      title={tooltip}
      style={style}
      className={toolbarClasses.className}
      ref={outerContainer}
      slot={slot}
      onClick={handleToolbarClick}
      {...rest}
    >
      <div className={classes.toolbar} data-component-name="ToolbarContent">
        {overflowNeeded &&
          React.Children.map(childrenWithRef, (item, index) => {
            if (index >= lastVisibleIndex + 1) {
              return React.cloneElement(item as ReactElement, { style: { visibility: 'hidden' } });
            }
            return item;
          })}
        {!overflowNeeded && childrenWithRef}
      </div>
      {overflowNeeded && (
        <div
          className={classes.overflowButtonContainer}
          title={i18nBundle.getText(SHOW_MORE)}
          data-component-name="ToolbarOverflowButtonContainer"
        >
          <OverflowPopover lastVisibleIndex={lastVisibleIndex} contentClass={classes.popoverContent}>
            {React.Children.toArray(children).map((child) => {
              if ((child as ReactElement).type === React.Fragment) {
                return (child as ReactElement).props.children;
              }
              return child;
            })}
          </OverflowPopover>
        </div>
      )}
    </CustomTag>
  );
});

Toolbar.defaultProps = {
  as: 'div',
  toolbarStyle: ToolbarStyle.Standard,
  design: ToolbarDesign.Auto,
  active: false
};

Toolbar.displayName = 'Toolbar';
export { Toolbar };
