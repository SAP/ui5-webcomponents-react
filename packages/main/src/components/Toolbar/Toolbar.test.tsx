import { fireEvent, render, screen } from '@shared/tests';
import { createChangeTagNameTest } from '@shared/tests/utils';
import { Text } from '@ui5/webcomponents-react/dist/Text';
import { Toolbar } from '@ui5/webcomponents-react/dist/Toolbar';
import { ToolbarDesign } from '@ui5/webcomponents-react/dist/ToolbarDesign';
import { ToolbarSeparator } from '@ui5/webcomponents-react/dist/ToolbarSeparator';
import { ToolbarSpacer } from '@ui5/webcomponents-react/dist/ToolbarSpacer';
import { ToolbarStyle } from '@ui5/webcomponents-react/dist/ToolbarStyle';
import React from 'react';

describe('Toolbar', () => {
  test('Renders with default Props', () => {
    expect(render(<Toolbar />).asFragment()).toMatchSnapshot();
  });

  test('Renders with children', () => {
    expect(
      render(
        <Toolbar active data-testid="toolbar">
          <Text>Item1</Text>
          <Text>Item2</Text>
          <Text>Item3</Text>
        </Toolbar>
      ).asFragment()
    ).toMatchSnapshot();
    expect(screen.getByText('Item1').textContent).toEqual('Item1');
    expect(screen.getByTestId('toolbar')).toHaveClass('Toolbar-active');
  });

  test('Renders with children as react fragments', () => {
    expect(
      render(
        <Toolbar active data-testid="toolbar">
          <>
            <Text>Item1</Text>
            <Text>Item2</Text>
            <Text>Item3</Text>
          </>
        </Toolbar>
      ).asFragment()
    ).toMatchSnapshot();
    expect(screen.getByText('Item1').textContent).toEqual('Item1');
    expect(screen.getByTestId('toolbar')).toHaveClass('Toolbar-active');
  });

  test('ToolbarSpacer', () => {
    const { getByTestId, asFragment } = render(
      <Toolbar data-testid="toolbar">
        <Text>Item1</Text>
        <ToolbarSpacer />
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    const toolbarSpacer = getByTestId('toolbar').children[0].children[1];
    expect(toolbarSpacer).toHaveClass('spacer');
    expect(toolbarSpacer).toHaveStyle('flex-grow: 1');

    expect(asFragment()).toMatchSnapshot();
  });

  test('ToolbarSeparator', () => {
    const { getByLabelText, asFragment } = render(
      <Toolbar data-testid="toolbar">
        <Text>Item1</Text>
        <ToolbarSeparator />
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );

    expect(getByLabelText('Separator')).toHaveClass('ToolbarSeparator-separator');

    expect(asFragment()).toMatchSnapshot();
  });

  test('overflow menu', () => {
    jest.spyOn(window, 'requestAnimationFrame').mockImplementation((cb) => cb());

    HTMLElement.prototype.getBoundingClientRect = jest.fn(function () {
      return {
        width: parseFloat(getComputedStyle(this).width || 200),
        height: 10,
        top: 0,
        left: 0,
        bottom: 0,
        right: 0
      };
    });
    const onOverflowChange = jest.fn();
    const { getByTitle, getAllByTestId, getAllByText, rerender, queryByTitle, getByText, getAllByLabelText } = render(
      <Toolbar data-testid="toolbar" style={{ width: '300px' }} onOverflowChange={onOverflowChange}>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item1
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item2
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item3
        </Text>
      </Toolbar>
    );
    expect(getByTitle('Show More')).toBeInTheDocument();
    expect(getAllByTestId('toolbar-item')).toHaveLength(5);
    expect(onOverflowChange).toHaveBeenCalledTimes(1);

    const item1 = getAllByText('Item1');
    const item2 = getAllByText('Item2');
    const item3 = getAllByText('Item3');
    expect(item1).toHaveLength(1);
    expect(item2).toHaveLength(2);
    expect(item3).toHaveLength(2);
    expect(item1[0]).not.toHaveStyle(`visibility: hidden`);
    expect(item2[0]).toHaveStyle(`visibility: hidden`);
    expect(item3[0]).toHaveStyle(`visibility: hidden`);

    expect(document.body).toMatchSnapshot();

    rerender(
      <Toolbar data-testid="toolbar" style={{ width: '1000px' }} onOverflowChange={onOverflowChange}>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item1
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item2
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item3
        </Text>
      </Toolbar>
    );
    const updatedItem1 = getByText('Item1');
    const updatedItem2 = getByText('Item2');
    const updatedItem3 = getByText('Item3');
    expect(queryByTitle('Show More')).toBeNull();
    expect(updatedItem1).toBeInTheDocument();
    expect(updatedItem2).toBeInTheDocument();
    expect(updatedItem3).toBeInTheDocument();
    expect(updatedItem1).not.toHaveStyle(`visibility: hidden`);
    expect(updatedItem2).not.toHaveStyle(`visibility: hidden`);
    expect(updatedItem3).not.toHaveStyle(`visibility: hidden`);
    expect(onOverflowChange).toHaveBeenCalledTimes(2);

    //with fragments
    rerender(
      <Toolbar data-testid="toolbar" style={{ width: '300px' }}>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item1
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item2
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item3
        </Text>
      </Toolbar>
    );
    expect(getByTitle('Show More')).toBeInTheDocument();
    expect(getAllByTestId('toolbar-item')).toHaveLength(5);

    const item1frag = getAllByText('Item1');
    const item2frag = getAllByText('Item2');
    const item3frag = getAllByText('Item3');
    expect(item1frag).toHaveLength(1);
    expect(item2frag).toHaveLength(2);
    expect(item3frag).toHaveLength(2);
    expect(item1frag[0]).not.toHaveStyle(`visibility: hidden`);
    expect(item2frag[0]).toHaveStyle(`visibility: hidden`);
    expect(item3frag[0]).toHaveStyle(`visibility: hidden`);

    expect(document.body).toMatchSnapshot();

    rerender(
      <Toolbar data-testid="toolbar" style={{ width: '1000px' }}>
        <>
          <Text data-testid="toolbar-item" style={{ width: '200px' }}>
            Item1
          </Text>
          <Text data-testid="toolbar-item" style={{ width: '200px' }}>
            Item2
          </Text>
          <Text data-testid="toolbar-item" style={{ width: '200px' }}>
            Item3
          </Text>
        </>
      </Toolbar>
    );

    const updatedItem1frag = getByText('Item1');
    const updatedItem2frag = getByText('Item2');
    const updatedItem3frag = getByText('Item3');
    expect(queryByTitle('Show More')).toBeNull();
    expect(updatedItem1frag).toBeInTheDocument();
    expect(updatedItem2frag).toBeInTheDocument();
    expect(updatedItem3frag).toBeInTheDocument();
    expect(updatedItem1frag).not.toHaveStyle(`visibility: hidden`);
    expect(updatedItem2frag).not.toHaveStyle(`visibility: hidden`);
    expect(updatedItem3frag).not.toHaveStyle(`visibility: hidden`);

    //with separator
    rerender(
      <Toolbar data-testid="toolbar" style={{ width: '300px' }}>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item1
        </Text>
        <ToolbarSeparator />
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item2
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '200px' }}>
          Item3
        </Text>
      </Toolbar>
    );

    const popoverSeparator = getAllByLabelText('Separator')[1];

    expect(popoverSeparator).toHaveClass('ToolbarSeparator-separator');
    expect(popoverSeparator).toHaveStyle('height: 0.0625rem; margin: 0.375rem 0.1875rem; width: 100%;');

    expect(document.body).toMatchSnapshot();
  });

  test('active', () => {
    const onClick = jest.fn();
    const { rerender } = render(
      <Toolbar data-testid="toolbar" onClick={onClick}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );

    fireEvent.click(screen.getByTestId('toolbar'));

    expect(onClick).not.toHaveBeenCalled();

    rerender(
      <Toolbar data-testid="toolbar" onClick={onClick} active>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );

    fireEvent.click(screen.getByTestId('toolbar'));

    expect(onClick).toHaveBeenCalled();
  });

  test('style', () => {
    const { rerender } = render(
      <Toolbar data-testid="toolbar">
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTestId('toolbar')).not.toHaveClass('Toolbar-clear');

    rerender(
      <Toolbar data-testid="toolbar" toolbarStyle={ToolbarStyle.Clear}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );
    expect(screen.getByTestId('toolbar')).toHaveClass('Toolbar-clear');
  });

  test.each([
    [ToolbarDesign.Info, 'Toolbar-info'],
    [ToolbarDesign.Solid, 'Toolbar-solid'],
    [ToolbarDesign.Transparent, 'Toolbar-transparent']
  ])('design: %s', (design, expected) => {
    render(
      <Toolbar data-testid={'toolbar'} design={design}>
        <Text>Item1</Text>
        <Text>Item2</Text>
        <Text>Item3</Text>
      </Toolbar>
    );

    expect(screen.getByTestId('toolbar')).toHaveClass(expected);
  });

  test('always visible items', () => {
    jest.spyOn(window, 'requestAnimationFrame').mockImplementation((cb) => cb());

    HTMLElement.prototype.getBoundingClientRect = jest.fn(function () {
      return {
        width: parseFloat(getComputedStyle(this).width || 100),
        height: 10,
        top: 0,
        left: 200,
        bottom: 0,
        right: 0
      };
    });
    const onOverflowChange = jest.fn();
    const { getByTitle, getAllByTestId, getAllByText, rerender, queryByTitle, getByText, getAllByLabelText } = render(
      <Toolbar
        data-testid="toolbar"
        style={{ width: '50px' }}
        onOverflowChange={onOverflowChange}
        numberOfAlwaysVisibleItems={2}
      >
        <Text data-testid="toolbar-item" style={{ width: '100px' }}>
          Item1
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '100px' }}>
          Item2
        </Text>
        <Text data-testid="toolbar-item" style={{ width: '100px' }}>
          Item3
        </Text>
      </Toolbar>
    );

    expect(getAllByText('Item1')).toHaveLength(1);
    expect(getAllByText('Item2')).toHaveLength(1);
    expect(getAllByText('Item3')).toHaveLength(2);
    expect(getAllByText('Item1')[0]).not.toHaveStyle(`visibility: hidden`);
    expect(getAllByText('Item2')[0]).not.toHaveStyle(`visibility: hidden`);
    expect(getAllByText('Item3')[0]).toHaveStyle(`visibility: hidden`);
  });

  createChangeTagNameTest(Toolbar);
});
