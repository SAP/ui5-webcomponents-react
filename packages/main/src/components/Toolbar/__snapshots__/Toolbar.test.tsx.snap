// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Renders with children 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer Toolbar-active"
    data-testid="toolbar"
  >
    <div
      class="Toolbar-toolbar"
      data-component-name="ToolbarContent"
    >
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item1
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item2
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar Renders with children as react fragments 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer Toolbar-active"
    data-testid="toolbar"
  >
    <div
      class="Toolbar-toolbar"
      data-component-name="ToolbarContent"
    >
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item1
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item2
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar Renders with default Props 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer"
  >
    <div
      class="Toolbar-toolbar"
      data-component-name="ToolbarContent"
    />
  </div>
</DocumentFragment>
`;

exports[`Toolbar ToolbarSeparator 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer"
    data-testid="toolbar"
  >
    <div
      class="Toolbar-toolbar"
      data-component-name="ToolbarContent"
    >
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item1
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <div
          aria-label="Separator"
          class="ToolbarSeparator-separator"
        />
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item2
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar ToolbarSpacer 1`] = `
<DocumentFragment>
  <div
    class="Toolbar-outerContainer"
    data-testid="toolbar"
  >
    <div
      class="Toolbar-toolbar"
      data-component-name="ToolbarContent"
    >
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item1
        </span>
      </div>
      <span
        class="spacer"
        style="flex-grow: 1;"
      />
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item2
        </span>
      </div>
      <div
        class="Toolbar-childContainer"
        data-component-name="ToolbarChildContainer"
      >
        <span
          class="Text-text"
        >
          Item3
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Toolbar overflow menu 1`] = `
<body>
  <div>
    <div
      class="Toolbar-outerContainer Toolbar-hasOverflow"
      data-testid="toolbar"
      style="width: 300px;"
    >
      <div
        class="Toolbar-toolbar"
        data-component-name="ToolbarContent"
      >
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item1
          </span>
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item2
          </span>
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item3
          </span>
        </div>
      </div>
      <div
        class="Toolbar-overflowButtonContainer"
        data-component-name="ToolbarOverflowButtonContainer"
        title="Show More"
      >
        <ui5-toggle-button
          design="Transparent"
          icon="overflow"
          ui5-toggle-button=""
        />
      </div>
    </div>
  </div>
  <ui5-popover
    horizontal-align="Center"
    placement-type="Bottom"
    ui5-popover=""
    vertical-align="Center"
  >
    <div
      class="Toolbar-popoverContent"
    >
      <span
        class="Text-text"
        data-testid="toolbar-item"
        style="width: 200px;"
      >
        Item2
      </span>
      <span
        class="Text-text"
        data-testid="toolbar-item"
        style="width: 200px;"
      >
        Item3
      </span>
    </div>
  </ui5-popover>
</body>
`;

exports[`Toolbar overflow menu 2`] = `
<body>
  <div>
    <div
      class="Toolbar-outerContainer Toolbar-hasOverflow"
      data-testid="toolbar"
      style="width: 300px;"
    >
      <div
        class="Toolbar-toolbar"
        data-component-name="ToolbarContent"
      >
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item1
          </span>
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item2
          </span>
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item3
          </span>
        </div>
      </div>
      <div
        class="Toolbar-overflowButtonContainer"
        data-component-name="ToolbarOverflowButtonContainer"
        title="Show More"
      >
        <ui5-toggle-button
          design="Transparent"
          icon="overflow"
          ui5-toggle-button=""
        />
      </div>
    </div>
  </div>
  <ui5-popover
    horizontal-align="Center"
    placement-type="Bottom"
    ui5-popover=""
    vertical-align="Center"
  >
    <div
      class="Toolbar-popoverContent"
    >
      <span
        class="Text-text"
        data-testid="toolbar-item"
        style="width: 200px;"
      >
        Item2
      </span>
      <span
        class="Text-text"
        data-testid="toolbar-item"
        style="width: 200px;"
      >
        Item3
      </span>
    </div>
  </ui5-popover>
</body>
`;

exports[`Toolbar overflow menu 3`] = `
<body>
  <div>
    <div
      class="Toolbar-outerContainer Toolbar-hasOverflow"
      data-testid="toolbar"
      style="width: 300px;"
    >
      <div
        class="Toolbar-toolbar"
        data-component-name="ToolbarContent"
      >
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item1
          </span>
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <div
            aria-label="Separator"
            class="ToolbarSeparator-separator"
          />
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item2
          </span>
        </div>
        <div
          class="Toolbar-childContainer"
          data-component-name="ToolbarChildContainer"
          style="visibility: hidden;"
        >
          <span
            class="Text-text"
            data-testid="toolbar-item"
            style="width: 200px;"
          >
            Item3
          </span>
        </div>
      </div>
      <div
        class="Toolbar-overflowButtonContainer"
        data-component-name="ToolbarOverflowButtonContainer"
        title="Show More"
      >
        <ui5-toggle-button
          design="Transparent"
          icon="overflow"
          ui5-toggle-button=""
        />
      </div>
    </div>
  </div>
  <ui5-popover
    horizontal-align="Center"
    placement-type="Bottom"
    ui5-popover=""
    vertical-align="Center"
  >
    <div
      class="Toolbar-popoverContent"
    >
      <div
        aria-label="Separator"
        class="ToolbarSeparator-separator"
        style="height: 0.0625rem; margin: 0.375rem 0.1875rem; width: 100%;"
      />
      <span
        class="Text-text"
        data-testid="toolbar-item"
        style="width: 200px;"
      >
        Item2
      </span>
      <span
        class="Text-text"
        data-testid="toolbar-item"
        style="width: 200px;"
      >
        Item3
      </span>
    </div>
  </ui5-popover>
</body>
`;
